src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs

```diff
@@ -357,85 +322,92 @@ void OutputXml (string iLLinkTrimXmlFilePath, string outputFileName)
 			XmlNode assemblyNode = linkerNode["assembly"];
 
 			foreach (string typeName in classNamesToClassMembers.Keys) {
-				XmlNode typeNode = doc.CreateElement ("type");
-				XmlAttribute typeFullName = doc.CreateAttribute ("fullname");
-				typeFullName.Value = typeName;
-				typeNode.Attributes.Append (typeFullName);
-
-				ClassMembers members = classNamesToClassMembers[typeName];
-
-				// We need to keep everyting in System.Runtime.InteropServices.WindowsRuntime and
-				// System.Threading.Volatile.
-				if (!typeName.StartsWith ("System.Runtime.InteropServices.WindowsRuntime") &&
-					!typeName.StartsWith ("System.Threading.Volatile")) {
-					if (members.keepAllFields) {
-						XmlAttribute preserve = doc.CreateAttribute ("preserve");
-						preserve.Value = "fields";
-						typeNode.Attributes.Append (preserve);
-					} else if ((members.fields == null) && (members.methods == null)) {
-						XmlAttribute preserve = doc.CreateAttribute ("preserve");
-						preserve.Value = "nothing";
-						typeNode.Attributes.Append (preserve);
-					}
+				AddXmlTypeNode (doc, assemblyNode, typeName, classNamesToClassMembers);
+			}
+			doc.Save (outputFileName);
+		}
 
-					if (!members.keepAllFields && (members.fields != null)) {
-						foreach (string field in members.fields) {
-							XmlNode fieldNode = doc.CreateElement ("field");
-							XmlAttribute fieldName = doc.CreateAttribute ("name");
-							fieldName.Value = field;
-							fieldNode.Attributes.Append (fieldName);
-							typeNode.AppendChild (fieldNode);
-						}
+		static void AddXmlTypeNode (XmlDocument doc, XmlNode assemblyNode, string typeName, Dictionary<string, ClassMembers> classToClassMems)
```

Comment:
Why not just pass `ClassMembers` - we already pass the type name, so the additional lookup is unnecessary (and the callers all do a foreach so they can directly specify the value without lookup as well.